mtt/src/mtt.rs

```diff
@@ -0,0 +1,688 @@
+    /// The address isn't aligned on a 2MB boundary
+    InvalidAligment2M,
+    /// The range type and length are mismatched
+    RangeTypeMismatch,
+    /// The specificied page size is yet supported
+    UnsupportedPageSize,
+    /// The caller specified an invalid number of pages for the range (example: 0,
+    /// or would overflow address space)
+    InvalidAddressRange,
+    /// The platform initialization hasn't been completed. The implementation is expected
+    /// to initialize pointers to the MTT L2 and L1 page pool using platform-FW tables.
+    PlatformInitNotCompleted,
+    /// The platform configuration for the MTT L2 and L1 is invalid
+    InvalidPlatformConfiguration,
+    /// The platform didn't allocate sufficient L1 page pool pages to map the specified address
+    InsufficientL1Pages,
+    /// L1 pages must be confidential
+    NonConfidentialL1Page,
+}
+
+/// Holds the result of a MTT operation.
+pub type Result<T> = core::result::Result<T, Error>;
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+/// API-related enums for non-confidential and confidential types
+pub enum MttMemoryType {
+    /// A confidential memory range
+    Confidential,
+    /// A non-confidential memory range
+    NonConfidential,
+}
+
+mod mtt_const {
+    pub const L2_TYPE_SHIFT: u64 = 34;
+    // Bits 63:36 must be zero
+    pub const L2_ZERO_MASK: u64 = 0xffff_fff0_0000_0000;
+    // Bits 35:34
+    pub const L2_TYPE_MASK: u64 = 0x000c_0000_0000;
+    // Bits 33:0
+    pub const L2_INFO_MASK: u64 = 0x0003_ffff_ffff;
+    // Bits 35:34 for a 1G non-confidential region
+    // By convention, this is a total of 16 identical consecutive entries, each mapping 64MB
+    pub const L2_1G_NC_TYPE: u64 = 0;
+    // Bits 35:34 for a 1G confidential region
+    // By convention, this is a total of 16 identical consecutive entries, each mapping 64MB
+    pub const L2_1G_C_TYPE: u64 = 1;
+    // Bits 35:34 for a 64-MB region mapped using a bit-mask in a 4K L1-page
+    pub const L2_L1_TYPE: u64 = 2;
+    // Bits 35:34 for a 64-MB region mapped using a 2MB sub-region bitmask
+    pub const L2_64M_TYPE: u64 = 3;
+    // L2 has 1MB entries
+    pub const MTT_L2_ENTRY_COUNT: usize = 1024 * 1024;
+    // L2 is 8MB long
+    pub const L2_BYTE_SIZE: usize = 1024 * 1024 * 8;
+    // A L1 page consists of 512 8-byte entries
+    pub const L1_ENTRY_COUNT: usize = 512;
+    // Bits 25:17 for the 9-bit index
+    pub const L1_INDEX_MASK: u64 = 0x03fe_0000;
+    pub const L1_INDEX_SHIFT: u64 = 17;
+    // Bits 16:12 for the 5-bit sub-index
+    pub const L1_SUBINDEX_MASK: u64 = 0x1_f000;
+    // 1-bit left shift for the 2-bits per entry for bits (16:12)
+    pub const L1_SUBINDEX_SHIFT: u64 = 11;
+    // 2-bits per entry
+    pub const L1_ENTRY_MASK: u64 = 0x3;
+}
+
+use mtt_const::*;
+
+struct Mtt1GEntry<'a, T>
+where
+    T: MemoryType,
+{
+    // 16-entries of the same type each spanning 64MB by convention.
+    entries: &'a mut [u64],
+    conf_state: PhantomData<T>,
```

Comment:
given the way the code has evolved, i'm not sure tagging the `Mtt1GEntry` type with the confidentiality of the range adds much anymore.